### ISSUE TITLE
Segfault of exiv2 with pr flag

### ISSUE THREAD

**Issue by LockOne** [Exiv2/exiv2#1097]
**Describe the bug**
Hi, we are currently working on fuzzing to automatically generate test cases, and we found a crash of exiv2 with pr flag.

**To Reproduce**
Running following command lead to crash:
`./exiv2 pr <attatched file>`
[exiv2_pr_crash.txt](https://github.com/Exiv2/exiv2/files/4184184/exiv2_pr_crash.txt)

I used 0.27.2 release version of exiv2.

**Desktop (please complete the following information):**
Ubuntu 16.04, clang 4.0.0

**Additional context**
This is call stack of the crashed execution:

```
#0  0x0000555555db5bd9 in Exiv2::Slice<unsigned char const*>::Slice(unsigned char const*, unsigned long, unsigned long) ()
#1  0x0000555555de05d3 in Exiv2::Slice<unsigned char const*> Exiv2::makeSliceUntil<unsigned char const>(unsigned char const*, unsigned long) ()
#2  0x0000555555877a84 in Exiv2::getUShort(unsigned char const*, Exiv2::ByteOrder) ()
#3  0x0000555555a4707c in Exiv2::Internal::CiffComponent::doRead(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#4  0x0000555555a4e449 in Exiv2::Internal::CiffComponent::read(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#5  0x0000555555a4d879 in Exiv2::Internal::CiffDirectory::readDirectory(unsigned char const*, unsigned int, Exiv2::ByteOrder) ()
#6  0x0000555555a4b796 in Exiv2::Internal::CiffDirectory::doRead(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#7  0x0000555555a4e449 in Exiv2::Internal::CiffComponent::read(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#8  0x0000555555a4d879 in Exiv2::Internal::CiffDirectory::readDirectory(unsigned char const*, unsigned int, Exiv2::ByteOrder) ()
#9  0x0000555555a4b796 in Exiv2::Internal::CiffDirectory::doRead(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#10 0x0000555555a4e449 in Exiv2::Internal::CiffComponent::read(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
...
#87306 0x0000555555a4b796 in Exiv2::Internal::CiffDirectory::doRead(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#87307 0x0000555555a4e449 in Exiv2::Internal::CiffComponent::read(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#87308 0x0000555555a4d879 in Exiv2::Internal::CiffDirectory::readDirectory(unsigned char const*, unsigned int, Exiv2::ByteOrder) ()
#87309 0x0000555555a4b796 in Exiv2::Internal::CiffDirectory::doRead(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#87310 0x0000555555a4e449 in Exiv2::Internal::CiffComponent::read(unsigned char const*, unsigned int, unsigned int, Exiv2::ByteOrder) ()
#87311 0x0000555555a4d879 in Exiv2::Internal::CiffDirectory::readDirectory(unsigned char const*, unsigned int, Exiv2::ByteOrder) ()
#87312 0x0000555555a646d6 in Exiv2::Internal::CiffHeader::read(unsigned char const*, unsigned int) ()
#87313 0x00005555559dd417 in Exiv2::CrwParser::decode(Exiv2::CrwImage*, unsigned char const*, unsigned int) ()
#87314 0x00005555559d95d1 in Exiv2::CrwImage::readMetadata() ()
#87315 0x0000555555646c65 in Action::Print::printSummary() ()
#87316 0x000055555564600d in Action::Print::run(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) ()
#87317 0x000055555562660b in main ()

```
Thanks.

**Comment by kevinbackhouse** [comment]
It's an infinite recursion because it keeps calling CiffDirectory::readDirectory() to parse the data at offset 0. The code is not enforcing that the directories should be nested.

I don't know the rules of the format, but my guess is that we should enforce that `offset_ >= start + 6` at [line 247](https://github.com/Exiv2/exiv2/blob/master/src/crwimage_int.cpp#L247), to enforce that the sub-directory is nested correctly. I'll post a PR for people to comment on.

This bug causes a crash due to stack exhaustion. There is no memory corruption.

**Comment by kevinbackhouse** [comment]
It turns out that the fix that I suggested is too simple, because the offset is sometimes at a lower address in memory. [This file](https://github.com/Exiv2/exiv2/blob/d4d4d766e9ade2376115eb41cc478eb195df1b39/test/data/exiv2-canon-powershot-s40.crw) is a counterexample. I have posted #1098 with the updated logic.

**Comment by clanmills** [comment]
I'm going to reopen this for Exiv2 0.27.3 on branch 0.27-maintenance.

**Comment by clanmills** [comment]
Closed on Exiv2 v0.27.3 via #1115
